lib/theory/src/Theory/Constraint/System/Graph/GraphRepr.hs

```diff
@@ -92,4 +108,149 @@ toEdgeList repr =
+
+-- Function to find the connected components within a cluster
+findConnectedComponents :: [Node] -> [Edge] -> [[Node]]
+findConnectedComponents nodes edges = go nodes []
+  where
+    -- Recursive function to find all nodes connected from a given node
+    expandCluster :: Node -> S.Set Th.NodeId -> [Node] -> [Edge] -> S.Set Th.NodeId
+    expandCluster node visited allNodes allEdges =
+      let nodeId = get nNodeId node
+          connectedNodes = [ tgt | SystemEdge ((src, _), (tgt, _)) <- allEdges, src == nodeId, tgt `S.notMember` visited ] ++
+                           [ src | SystemEdge ((src, _), (tgt, _)) <- allEdges, tgt == nodeId, src `S.notMember` visited ]
+          newVisited = S.insert nodeId visited
+      in foldr (\nid acc -> if nid `S.member` visited then acc else expandCluster (findNodeById nid allNodes) newVisited allNodes allEdges `S.union` acc) (S.singleton nodeId) connectedNodes
+
+    findNodeById :: Th.NodeId -> [Node] -> Node
+    findNodeById nodeId allNodes = head $ filter (\n -> get nNodeId n == nodeId) allNodes
+
+    -- Main function to find all connected components
+    go :: [Node] -> [[Node]] -> [[Node]]
+    go [] components = components
+    go (n:ns) components =
+      let componentIds = S.toList $ expandCluster n S.empty (n:ns) edges
+          component = filter (\node -> get nNodeId node `elem` componentIds) (n:ns)
+          remainingNodes = filter (`notElem` component) ns
+      in go remainingNodes (component : components)
+
+
+-- Create the sub-clusters of an agent and add them to GraphRepr
+addSubClustersByAgent :: GraphRepr -> GraphRepr
+addSubClustersByAgent repr =
+    let nodesByAgent = groupNodesByAgent (get grNodes repr)
+        edges = get grEdges repr
+        createSubClusters agent nodes =
+            let connectedComponents = findConnectedComponents nodes (filterEdgesForCluster nodes edges)
+            in zipWith (\i component -> createCluster (agent ++ "_Session_" ++ show (i :: Integer)) component (filterEdgesForCluster component edges)) [1..] connectedComponents
+        subClusters = concatMap (\(agent, nodes) -> createSubClusters agent nodes) (Map.toList nodesByAgent)
+        clusterEdges = concatMap (get cEdges) subClusters
+        clusteredNodes = concatMap (get cNodes) subClusters
+        remainingEdges = filter (`notElem` clusterEdges) edges
+        remainingNodes = filter (`notElem` clusteredNodes) (get grNodes repr)
+    in set grClusters subClusters $
+       set grEdges remainingEdges $
+       set grNodes remainingNodes repr
+
+
+
+----------------------------------------------------
+-- Clustering based on the name of the rules.
+----------------------------------------------------
+
+-- Function to add intelligent clusters with similar rule names to GraphRepr
+addIntelligentClusterWithSubClusters :: GraphRepr -> GraphRepr
+addIntelligentClusterWithSubClusters repr =
```

Comment:
addIntelligentClusterWithSubClusters and addSubClustersByAgent seem to identical except for the first line. Why not factor that out, to avoid copy-paste?